### PR TITLE
Add arrays_overlap Presto function

### PR DESCRIPTION
**Pull request by karteekmurthys**

### CODE REVIEW

File: velox/functions/prestosql/ArraysOverlap.cpp [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+  return VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH(
+      createTyped, elementType->kind(), inputArgs);
+}
+
+std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+  // array(T), array(T) -> array(T)
+  return {exec::FunctionSignatureBuilder()
+              .typeVariable("T")
+              .returnType("array(T)")
```
Comment (aditi-pandit) [inline velox/functions/prestosql/ArraysOverlap.cpp:148]: Shouldn't this be array of boolean ?

File: velox/functions/lib/LambdaFunctionUtil.cpp [hunk 2]
```diff
@@ -154,4 +154,46 @@ MapVectorPtr flattenMap(
...
+  // Decode and acquire array elements vector.
+  auto elementsVector = baseArrayVector->elements();
+  auto elementsSelectivityRows = toElementRows(
+      elementsVector->size(), rows, baseArrayVector, decodedVector->indices());
+  elementsDecoder.get()->decode(*elementsVector, elementsSelectivityRows);
+  auto decodedElementsVector = elementsDecoder.get();
+  return decodedElementsVector;
+}
+
+void validateType(
```
Comment (aditi-pandit) [inline velox/functions/lib/LambdaFunctionUtil.cpp:173]: Make this more specific : validateMatchingArrayTypes.

File: velox/functions/lib/LambdaFunctionUtil.cpp [hunk 3]
```diff
@@ -154,4 +154,46 @@ MapVectorPtr flattenMap(
...
+    const std::vector<exec::VectorFunctionArg>& inputArgs,
+    const std::string name,
+    const vector_size_t expectedArgCount) {
+  VELOX_USER_CHECK_EQ(
+      inputArgs.size(),
+      expectedArgCount,
+      "{} requires exactly two parameters",
```
Comment (aditi-pandit) [inline velox/functions/lib/LambdaFunctionUtil.cpp:180]: This should be expectedArgCount instead of two.

File: velox/functions/prestosql/ArraysOverlap.cpp [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    };
+
+    if (constantSet_.has_value()) {
+      rows.applyToSelected(
+          [&](vector_size_t row) { processRow(row, *constantSet_); });
+    }
+    // General case when no arrays are constant and both sets need to be
+    // computed for each row.
+    else {
+      exec::LocalDecodedVector rightDecoder(context, *right, rows);
```
Comment (aditi-pandit) [inline velox/functions/prestosql/ArraysOverlap.cpp:82]: You can use getDecodedElementsFromArrayVector for right side also here I think. Similar for ArraysIntersectExcept.

File: velox/functions/prestosql/ArraysOverlap.cpp [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    exec::LocalDecodedVector decoder(context, *left, rows);
+    exec::LocalDecodedVector elementsDecoder(context);
+    auto decodedLeftElements =
+        getDecodedElementsFromArrayVector(decoder, elementsDecoder, rows);
+
+    FlatVector<bool>* resBoolVector = (*result)->asFlatVector<bool>();
+    auto processRow = [&](const vector_size_t rowId,
+                          const SetWithNull<T>& rightSet) {
+      auto decodedLeftArray = decoder.get();
+      auto baseLeftArray = decodedLeftArray->base()->as<ArrayVector>();
```
Comment (aditi-pandit) [inline velox/functions/prestosql/ArraysOverlap.cpp:54]: These 2 lines can be moved out of processRow. The local variables of the function can be accessed in the lambda as local variables.

File: velox/functions/prestosql/ArraysOverlap.cpp [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        }
+        if (rightSet.set.count(decodedLeftElements->valueAt<T>(i)) > 0) {
+          // Found and overlapping element. Add to result set.
+          resBoolVector->set(rowId, true);
+          return;
+        }
+      }
+      // If none of them is found in the rightSet, set false for current row
+      // indicating there are no overlapping elements with rightSet.
+      resBoolVector->set(rowId, false);
```
Comment (aditi-pandit) [inline velox/functions/prestosql/ArraysOverlap.cpp:72]: Doesn't look like this is returned ?

**Review by aditi-pandit — COMMENTED** [review]
Meta comment: There is something called Lambda functions in Presto. And I think LambdaFunctionsUtil is very specific to it. There are 2 cleaner options here:
i) Create a new ArraysUtil file.
ii) Move arrays_overlap to the same file as ArrayIntersectExcept. After the refactoring it doesn't seem like much new code will be added there.

File: velox/functions/prestosql/ArraysOverlap.cpp (hunk 5 above)
Comment (karteekmurthys) [reply velox/functions/prestosql/ArraysOverlap.cpp:54]: I moved extracting the decodedLeftArray and baseLeftArray as part of getDecodedElementsFromArrayVector.
That is the reason I need to re-fetch them here.

File: velox/functions/prestosql/ArraysOverlap.cpp (hunk 6 above)
Comment (karteekmurthys) [reply velox/functions/prestosql/ArraysOverlap.cpp:72]: resBoolVector is captured by the lambda.

File: velox/functions/sparksql/Register.cpp [hunk 7]
```diff
@@ -45,6 +45,7 @@ static void workAroundRegistrationMacro(const std::string& prefix) {
   VELOX_REGISTER_VECTOR_FUNCTION(udf_array_contains, prefix + "array_contains");
   VELOX_REGISTER_VECTOR_FUNCTION(
       udf_array_intersect, prefix + "array_intersect");
+  VELOX_REGISTER_VECTOR_FUNCTION(udf_arrays_overlap, prefix + "arrays_overlap");
```
Comment (mbasmanova) [inline velox/functions/sparksql/Register.cpp:48]: arrays_overlap is a Presto function, hence, should not appear in sparksql folder. Please, remove.

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 8]
```diff
@@ -65,6 +62,22 @@ void generateSet(
   }
 }
 
+DecodedVector* getDecodedElementsFromArrayVector(
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:65]: nit: this name is a bit long; how about decodeArrayElements?

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 9]
```diff
@@ -65,6 +62,22 @@ void generateSet(
...
+    exec::LocalDecodedVector& decoder,
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:66]: nit: decoder -> arrayDecoder for readability

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 10]
```diff
@@ -130,15 +143,9 @@ class ArrayIntersectExceptFunction : public exec::VectorFunction {
...
-    auto leftElementsVector = baseLeftArray->elements();
-    auto leftElementsRows = toElementRows(
-        leftElementsVector->size(),
-        rows,
-        baseLeftArray,
-        decodedLeftArray->indices());
-    exec::LocalDecodedVector leftElementsHolder(
-        context, *leftElementsVector, leftElementsRows);
-    auto decodedLeftElements = leftElementsHolder.get();
+    exec::LocalDecodedVector leftElementsDecoder(context);
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:146]: This looks like an unrelated refactoring. It would be nice to put it into a separate PR to allow the review to focus on material changes.

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 11]
```diff
@@ -226,24 +233,19 @@ class ArrayIntersectExceptFunction : public exec::VectorFunction {
...
+      auto decodedRightElements = getDecodedElementsFromArrayVector(
+          rightHolder, rightElementsHolder, rows);
       SetWithNull<T> rightSet;
 
       rows.applyToSelected([&](vector_size_t row) {
-        auto idx = decodedRightArray->index(row);
-        generateSet<T>(baseRightArray, decodedRightElements, idx, rightSet);
+        auto idx = rightHolder.get()->index(row);
+        generateSet<T>(
+            rightHolder.get()->base()->as<ArrayVector>(),
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:245]: "as" is a dynamic cast that tends to be expensive; it would be nice to keep it outside of the loop.

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 12]
```diff
@@ -269,29 +271,143 @@ class ArrayIntersectExceptFunction : public exec::VectorFunction {
 
   // If there's a `constantSet`, whether it refers to left or right-hand side.
   const bool isLeftConstant_{false};
-};
+}; // class ArrayIntersectExcept
+
+template <typename T>
+class ArraysOverlapFunction : public exec::VectorFunction {
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:277]: Does this function support all types or only scalar types for the array elements? If it only supports scalar types, please, update the documentation and PR description to clarify this limitation. How are you thinking about adding support for all types?

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 13]
```diff
@@ -269,29 +271,143 @@ class ArrayIntersectExceptFunction : public exec::VectorFunction {
...
+ public:
+  ArraysOverlapFunction() {}
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:279]: = default

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 14]
```diff
@@ -269,29 +271,143 @@ class ArrayIntersectExceptFunction : public exec::VectorFunction {
...
+
+  explicit ArraysOverlapFunction(
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:281]: explicit is only needed for a single-argument constructor; this constructor has 2 arguments, hence, explicit is not needed.

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 15]
```diff
@@ -269,29 +271,143 @@ class ArrayIntersectExceptFunction : public exec::VectorFunction {
...
+      bool isLeftConstant)
+      : constantSet_(std::move(constantSet)), isLeftConstant_(isLeftConstant) {}
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& /* outputType */,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    // if one of them is a ConstantVector.
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:292]: Comments should be full sentences, e.g. start with a capital letter and end with a period. Please, review and fix throughout the PR.

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 16]
```diff
@@ -269,29 +271,143 @@ class ArrayIntersectExceptFunction : public exec::VectorFunction {
...
+    if (constantSet_.has_value() && isLeftConstant_) {
+      std::swap(left, right);
+    }
+    exec::LocalDecodedVector decoder(context, *left, rows);
+    exec::LocalDecodedVector elementsDecoder(context);
+    auto decodedLeftElements =
+        getDecodedElementsFromArrayVector(decoder, elementsDecoder, rows);
+
+    BaseVector::ensureWritable(rows, BOOLEAN(), context->pool(), result);
+    auto resBoolVector = (*result)->template asFlatVector<bool>();
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:306]: do not abbreviate: resBoolVector -> resultBoolVector, but flatResult name might be clearer

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 17]
```diff
@@ -269,29 +271,143 @@ class ArrayIntersectExceptFunction : public exec::VectorFunction {
...
+    auto processRow = [&](const vector_size_t rowId,
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:307]: - rowId -> row to match the rest of the code base
- const for row is not needed
- use auto for brevity

```
[&] (auto row, const SetWithNull<T>& rightSet)
```

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 18]
```diff
@@ -269,29 +271,143 @@ class ArrayIntersectExceptFunction : public exec::VectorFunction {
...
+                          const SetWithNull<T>& rightSet) {
+      auto decodedLeftArray = decoder.get();
+      auto baseLeftArray = decodedLeftArray->base()->as<ArrayVector>();
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:310]: move this outside of the loop as dynamic cast tends to be expensive

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 19]
```diff
@@ -269,29 +271,143 @@ class ArrayIntersectExceptFunction : public exec::VectorFunction {
...
+      auto size = baseLeftArray->sizeAt(idx);
+
+      for (auto i = offset; i < (offset + size); ++i) {
+        // For each element in the current row search for it in the rightSet.
+        if (decodedLeftElements->isNullAt(i)) {
+          // Arrays overlap skips null values.
+          continue;
+        }
+        if (rightSet.set.count(decodedLeftElements->valueAt<T>(i)) > 0) {
+          // Found and overlapping element. Add to result set.
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:322]: typo: and -> an

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 20]
```diff
@@ -269,29 +271,143 @@ class ArrayIntersectExceptFunction : public exec::VectorFunction {
...
+          resBoolVector->set(rowId, true);
+          return;
+        }
+      }
+      // If none of them is found in the rightSet, set false for current row
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:327]: > Returns null if there are no non-null elements in common but either array contains null.

I don't see where this logic is implemented. Can you give me a pointer? Is there a test for that?

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 21]
```diff
@@ -269,29 +271,143 @@ class ArrayIntersectExceptFunction : public exec::VectorFunction {
...
-        "array_intersect function requires all arguments of the same type: {} vs. {}",
+        "{} function requires all arguments of the same type: {} vs. {}",
+        name,
         arg.type->toString(),
         arrayType->toString());
   }
 }
 
+template <typename T>
+void validateConstVecAndGenerateSet(
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:393]: do not abbreviate: validateConstantVectorAndGenerateSet

Based on usage of this function, it would be more convenient if it returned SetWithNull<T> instead of taking it as an output parameter.

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 22]
```diff
@@ -348,33 +448,68 @@ std::shared_ptr<exec::VectorFunction> createArrayIntersect(
...
+      createTypedArraysIntersectExcept,
+      /* isIntersect */ false,
+      elementType->kind(),
+      inputArgs);
 }
 
-std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+std::vector<std::shared_ptr<exec::FunctionSignature>> signatures(
+    const std::string returnType) {
   // array(T), array(T) -> array(T)
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:463]: this comment needs updating

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 23]
```diff
@@ -348,33 +448,68 @@ std::shared_ptr<exec::VectorFunction> createArrayIntersect(
...
+  auto elementType = inputArgs.front().type->childAt(0);
+
+  return VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH(
+      createTypedArraysOverlap, elementType->kind(), inputArgs);
+}
 } // namespace
 
+VELOX_DECLARE_STATEFUL_VECTOR_FUNCTION(
+    udf_arrays_overlap,
+    signatures("array(boolean)"),
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:503]: Should this be "boolean" and not "array(boolean)"?

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 24]
```diff
@@ -348,33 +448,68 @@ std::shared_ptr<exec::VectorFunction> createArrayIntersect(
...
+    const std::vector<exec::VectorFunctionArg>& inputArgs) {
+  VELOX_CHECK_EQ(inputArgs.size(), 2);
+  BaseVector* left = inputArgs[0].constantValue.get();
+  BaseVector* right = inputArgs[1].constantValue.get();
+  using T = typename TypeTraits<kind>::NativeType;
+  const bool isLeftConstant = (left != nullptr);
+  if (left == nullptr && right == nullptr) {
+    return std::make_shared<ArraysOverlapFunction<T>>();
+  }
+  BaseVector* baseVector = (left != nullptr) ? left : right;
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:483]: use isLeftConstant here: isLeftConstant ? left : right

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 25]
```diff
@@ -348,33 +448,68 @@ std::shared_ptr<exec::VectorFunction> createArrayIntersect(
...
 }
 
+template <TypeKind kind>
+const std::shared_ptr<exec::VectorFunction> createTypedArraysOverlap(
+    const std::vector<exec::VectorFunctionArg>& inputArgs) {
+  VELOX_CHECK_EQ(inputArgs.size(), 2);
+  BaseVector* left = inputArgs[0].constantValue.get();
+  BaseVector* right = inputArgs[1].constantValue.get();
+  using T = typename TypeTraits<kind>::NativeType;
+  const bool isLeftConstant = (left != nullptr);
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:479]: move this variable closer to where it is used, e.g. after the `if (left == nullptr && right == nullptr)` statement.

**Review by mbasmanova — COMMENTED** [review]
@karteekmurthys Karteek, thank you for adding arrays_overlap Presto function. It looks like this function doesn't support non-primitive types for array elements. If that's the case, it would be simpler to implement it as a simple function.

File: velox/functions/prestosql/ArrayIntersectExcept.cpp (hunk 10 above)
Comment (karteekmurthys) [reply velox/functions/prestosql/ArrayIntersectExcept.cpp:146]: @mbasmanova Can we retain this for now. Since I have already moved out common code and refactored other functions, there is too much refactoring to undo. Apologies, will make sure future PRs are concise.

File: velox/functions/prestosql/ArrayIntersectExcept.cpp (hunk 20 above)
Comment (karteekmurthys) [reply velox/functions/prestosql/ArrayIntersectExcept.cpp:327]: Yes, i missed that. Thanks for pointing out:

```
presto:default> select * from t1;
      c0      
--------------
 [1, 2, 3]    
 [5, 4, null] 
 [5, 4]       
(3 rows)

Query 20220216_000754_00015_pg2ci, FINISHED, 1 node
Splits: 17 total, 17 done (100.00%)
318ms [3 rows, 315B] [9 rows/s, 990B/s]

presto:default> select arrays_overlap(c0, array[1,2]) from t1;
 _col0 
-------
 true  
 NULL  
 false 
(3 rows)
```

This how prestoDB returns for non-overlapping case with null and without null.

File: velox/functions/prestosql/ArrayIntersectExcept.cpp (hunk 12 above)
Comment (karteekmurthys) [reply velox/functions/prestosql/ArrayIntersectExcept.cpp:273]: For now this only supports scalar types, similar to Array intersect and except functions. We have not thought about complex types support yet.

File: velox/functions/prestosql/ArrayIntersectExcept.cpp (hunk 20 above)
Comment (karteekmurthys) [reply velox/functions/prestosql/ArrayIntersectExcept.cpp:323]: I switched to using nulls to adhere to prestodb standard.

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 26]
```diff
@@ -269,29 +267,145 @@ class ArrayIntersectExceptFunction : public exec::VectorFunction {
...
+    }
+    exec::LocalDecodedVector arrayDecoder(context, *left, rows);
+    exec::LocalDecodedVector elementsDecoder(context);
+    auto decodedLeftElements =
+        decodeArrayElements(arrayDecoder, elementsDecoder, rows);
+
+    BaseVector::ensureWritable(rows, BOOLEAN(), context->pool(), result);
+    auto resultBoolVector = (*result)->template asFlatVector<bool>();
+    auto processRow = [&](auto row, const SetWithNull<T>& rightSet) {
+      auto decodedLeftArray = arrayDecoder.get();
```
Comment (aditi-pandit) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:301]: Move these 2 lines outside of the processRow as they are common.

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 27]
```diff
@@ -269,29 +267,145 @@ class ArrayIntersectExceptFunction : public exec::VectorFunction {
...
+    else {
+      exec::LocalDecodedVector rightDecoder(context, *right, rows);
+      exec::LocalDecodedVector rightElementsDecoder(context);
+      auto decodedRightElements =
+          decodeArrayElements(rightDecoder, rightElementsDecoder, rows);
+      SetWithNull<T> rightSet;
+
+      rows.applyToSelected([&](vector_size_t row) {
+        auto idx = rightDecoder.get()->index(row);
+        auto baseRightArray = rightDecoder.get()->base()->as<ArrayVector>();
```
Comment (aditi-pandit) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:344]: This line can be moved outside of the lambda.

File: velox/functions/prestosql/tests/ArraysOverlapTest.cpp [hunk 28]
```diff
@@ -0,0 +1,220 @@
...
+         {std::numeric_limits<T>::min(), 2.02, -2.001, 1},
+         {std::numeric_limits<T>::max(), 8.0001, std::nullopt},
+         {9.0009,
+          std::numeric_limits<T>::infinity(),
+          std::numeric_limits<T>::max()},
+         {std::numeric_limits<T>::quiet_NaN(), 9.0009},
+         {std::numeric_limits<T>::quiet_NaN(), 9.0009, std::nullopt}});
+    auto array2 = makeNullableArrayVector<T>(
+        {{1.0, -2.0, 4.0},
+         {std::numeric_limits<T>::min(), 2.0199, -2.001, 1.000001},
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/ArraysOverlapTest.cpp:67]: maybe change this case so that the only overlapping value is min() validating it matches.

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 29]
```diff
@@ -269,29 +267,145 @@ class ArrayIntersectExceptFunction : public exec::VectorFunction {
...
+      : constantSet_(std::move(constantSet)), isLeftConstant_(isLeftConstant) {}
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& /* outputType */,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    // If one of them is a ConstantVector, iterate through the other
+    // ArrayVector->elementsFlatVector and do constantSet_.find().
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:287]: Looks like there are some typos: ArrayVector->elementsFlatVector. Would you double check and fix as needed?

File: velox/functions/prestosql/ArrayIntersectExcept.cpp (hunk 26 above)
Comment (mbasmanova) [reply velox/functions/prestosql/ArrayIntersectExcept.cpp:301]: +1

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 30]
```diff
@@ -269,29 +267,145 @@ class ArrayIntersectExceptFunction : public exec::VectorFunction {
...
+  std::optional<SetWithNull<T>> constantSet_;
 
-void validateType(const std::vector<exec::VectorFunctionArg>& inputArgs) {
+  // If there's a `constantSet`, whether it refers to left or right-hand side.
+  const bool isLeftConstant_{false};
+}; // class ArraysOverlapFunction
+
+void validateMatchingArrayTypes(
+    const std::vector<exec::VectorFunctionArg>& inputArgs,
+    const std::string name,
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:363]: pass name by reference

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 31]
```diff
@@ -348,33 +446,67 @@ std::shared_ptr<exec::VectorFunction> createArrayIntersect(
...
+      createTypedArraysIntersectExcept,
+      /* isIntersect */ false,
+      elementType->kind(),
+      inputArgs);
 }
 
-std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
-  // array(T), array(T) -> array(T)
+std::vector<std::shared_ptr<exec::FunctionSignature>> signatures(
+    const std::string returnType) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:460]: pass returnType by reference

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 32]
```diff
@@ -348,33 +446,67 @@ std::shared_ptr<exec::VectorFunction> createArrayIntersect(
...
+    const std::vector<exec::VectorFunctionArg>& inputArgs) {
+  VELOX_CHECK_EQ(inputArgs.size(), 2);
+  BaseVector* left = inputArgs[0].constantValue.get();
+  BaseVector* right = inputArgs[1].constantValue.get();
+  using T = typename TypeTraits<kind>::NativeType;
+  if (left == nullptr && right == nullptr) {
+    return std::make_shared<ArraysOverlapFunction<T>>();
+  }
+  const bool isLeftConstant = (left != nullptr);
+  BaseVector* baseVector = isLeftConstant ? left : right;
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:480]: use "auto" here and throughout the diff

File: velox/functions/prestosql/tests/ArraysOverlapTest.cpp [hunk 33]
```diff
@@ -0,0 +1,220 @@
...
+        {1, 1, -2, -2, -2, 4, 8},
+    });
+    auto array2 = makeNullableArrayVector<T>({
+        {1, -2, 4},
+        {1, -2, 4},
+        {1, -2, 4},
+        {1, -2, 4},
+    });
+    auto expected =
+        makeNullableFlatVector<bool>({true, true, std::nullopt, true});
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArraysOverlapTest.cpp:49]: Can you also test values that produce "false" result?

File: velox/functions/prestosql/tests/ArraysOverlapTest.cpp [hunk 34]
```diff
@@ -0,0 +1,220 @@
...
+  }
+  template <typename T>
+  void testInt() {
+    auto array1 = makeNullableArrayVector<T>({
+        {1, -2, 3, std::nullopt, 4, 5, 6, std::nullopt},
+        {1, 2, -2, 1},
+        {3, 8, std::nullopt},
+        {1, 1, -2, -2, -2, 4, 8},
+    });
+    auto array2 = makeNullableArrayVector<T>({
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArraysOverlapTest.cpp:42]: Can you also test arrays with null for "array2" and null arrays for both "array1" and "array2"?

File: velox/functions/prestosql/tests/ArraysOverlapTest.cpp [hunk 35]
```diff
@@ -0,0 +1,220 @@
...
+         {std::numeric_limits<T>::quiet_NaN(), 9.0009, std::nullopt}});
+    auto array2 = makeNullableArrayVector<T>(
+        {{1.0, -2.0, 4.0},
+         {std::numeric_limits<T>::min(), 2.0199, -2.001, 1.000001},
+         {1.0001, -2.02, std::numeric_limits<T>::max(), 8.00099},
+         {9.0009, std::numeric_limits<T>::infinity()},
+         {9.0009, std::numeric_limits<T>::quiet_NaN()},
+         {9.0}});
+    auto expected = makeNullableFlatVector<bool>(
+        {true, true, true, true, true, std::nullopt});
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArraysOverlapTest.cpp:73]: same as above

File: velox/functions/prestosql/tests/ArraysOverlapTest.cpp [hunk 36]
```diff
@@ -0,0 +1,220 @@
...
+       {std::nullopt}});
+
+  auto expected = makeNullableFlatVector<bool>(
+      {true, true, true, false, true, true, true, true, std::nullopt});
+
+  testExpr(expected, "arrays_overlap(C0, C1)", {array1, array2});
+  testExpr(expected, "arrays_overlap(C1, C0)", {array1, array2});
+}
+
+//// Test inline strings.
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArraysOverlapTest.cpp:122]: Use 3 slashes here and throughout the diff.

File: velox/functions/prestosql/tests/ArraysOverlapTest.cpp [hunk 37]
```diff
@@ -0,0 +1,220 @@
...
+TEST_F(ArraysOverlapTest, strArrays) {
+  using S = StringView;
+
+  auto array1 = makeNullableArrayVector<StringView>({
+      {S("a"), std::nullopt, S("b")},
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArraysOverlapTest.cpp:127]: consider "a"_sv

File: velox/functions/prestosql/tests/ArraysOverlapTest.cpp [hunk 38]
```diff
@@ -0,0 +1,220 @@
...
+
+  auto expected = makeNullableFlatVector<bool>(
+      {true, true, true, false, true, true, true, true, std::nullopt});
+
+  testExpr(expected, "arrays_overlap(C0, C1)", {array1, array2});
+  testExpr(expected, "arrays_overlap(C1, C0)", {array1, array2});
+}
+
+//// Test inline strings.
+TEST_F(ArraysOverlapTest, strArrays) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArraysOverlapTest.cpp:123]: do not abbreviate: strArrays -> shortStrings or inlineStrings

File: velox/functions/prestosql/tests/ArraysOverlapTest.cpp [hunk 39]
```diff
@@ -0,0 +1,220 @@
...
+      {S("abc"), S("a"), S("b")},
+  });
+  auto expected =
+      makeNullableFlatVector<bool>({true, true, std::nullopt, false});
+  testExpr(expected, "arrays_overlap(C0, C1)", {array1, array2});
+  testExpr(expected, "arrays_overlap(C1, C0)", {array1, array2});
+}
+
+// Test non-inline (> 12 length) strings.
+TEST_F(ArraysOverlapTest, longStrArrays) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArraysOverlapTest.cpp:145]: do not abbreviate: longStrArrays -> longStrings

**Review by mbasmanova — COMMENTED** [review]
@karteekmurthys Looks good to me % some comments below. Please, address Aditi's comments as well.

File: velox/functions/prestosql/tests/ArraysOverlapTest.cpp [hunk 40]
```diff
@@ -0,0 +1,237 @@
...
+      true,
+      std::nullopt,
+      std::nullopt,
+      true,
+  });
+  testExpr(expected, "arrays_overlap(C0, ARRAY[1,NULL,4])", {array1});
+  testExpr(expected, "arrays_overlap(ARRAY[1,NULL,4], C0)", {array1});
+}
+
+TEST_F(ArraysOverlapTest, wrongTypes) {
```
Comment (majetideepak) [inline velox/functions/prestosql/tests/ArraysOverlapTest.cpp:216]: should we add a test for array dimension mismatch between the input arrays?

File: velox/functions/prestosql/ArrayIntersectExcept.cpp [hunk 41]
```diff
@@ -65,6 +62,22 @@ void generateSet(
   }
 }
 
+DecodedVector* decodeArrayElements(
+    exec::LocalDecodedVector& arrayDecoder,
```
Comment (aditi-pandit) [inline velox/functions/prestosql/ArrayIntersectExcept.cpp:66]: Can this argument be const ?

**Review by aditi-pandit — APPROVED** [review]

File: velox/functions/prestosql/tests/ArraysOverlapTest.cpp (hunk 40 above)
Comment (karteekmurthys) [reply velox/functions/prestosql/tests/ArraysOverlapTest.cpp:216]: E20220304 10:59:56.535413 729935360 Exceptions.h:68] /Users/karteek/workspace/velox/./velox/vector/SelectivityVector.h87setValidRangeend <= bits_.size() * sizeof(bits_[0]) * 8(1128481602 vs. 64)RUNTIMEINVALID_STATE
I20220304 11:00:58.165135 729935360 Expr.cpp:263] Inside: 8 from 0 to 8 wrap FLAT expr arrays_overlap(c0, c1)
unknown file: Failure
C++ exception with description "Exception: VeloxRuntimeError
Error Source: RUNTIME

It doesn't allow us create arrays with different lengths. The call fails before reaching arrays_overlap.

File: velox/functions/prestosql/ArrayIntersectExcept.cpp (hunk 41 above)
Comment (karteekmurthys) [reply velox/functions/prestosql/ArrayIntersectExcept.cpp:66]: The arrayDecoder.get() call and elementsDecoder.get() calls are not const functions, so can't make them const.

**Review by mbasmanova — APPROVED** [review]
@karteekmurthys Looks good to me. Thanks.
